doc/manual/src/language/import-from-derivation.md

```diff
@@ -0,0 +1,86 @@
+[derivation]: @docroot@/glossary.md#gloss-derivation
+[realised]: @docroot@/glossary.md#gloss-realise
+
+This has performance implications:
+Since evaluation is sequential, each required store object that is not already in the store will also be realised sequentially.
+
+Passing an expression `expr` which evaluates to a store path to any built-in function that reads from the filesystem constitutes Import From Derivation:
+
+- [`import`](./builtins.md#builtins-import)` expr`
+- [`builtins.readFile`](./builtins.md#builtins-readFile)` expr`
+- [`builtins.readDir`](./builtins.md#builtins-readDir)` expr`
+- [`builtins.pathExists`](./builtins.md#builtins-pathExists)` expr`
+- [`builtins.filterSource`](./builtins.md#builtins-filterSource)` f expr`
+- [`builtins.path`](./builtins.md#builtins-path)` { path = expr; }`
+- [`builtins.hashFile`](./builtins.md#builtins-hashFile)` t expr`
+- `builtins.scopedImport x drv`
+
+Realising store objects during evaluation can be disabled by setting [`allow-import-from-derivation`](../command-ref/conf-file.md#conf-allow-import-from-derivation) to `false`.
+
+## Example
+
+In the following Nix expression, the inner derivation `drv` produces a file containing `"hello"`.
+
+```nix
+# IFD.nix
+let
+  drv = derivation {
+    name = "hello";
+    builder = /bin/sh;
+    args = [ "-c" ''echo \"hello\" > $out'' ];
+    system = builtins.currentSystem;
+  };
+in "${import drv} world"
+```
+
+```shellSession
+nix-instantiate IFD.nix --eval --read-write-mode
+```
+
+```
+building '/nix/store/348q1cal6sdgfxs8zqi9v8llrsn4kqkq-hello.drv'...
+"hello world"
+```
+
+Since `"hello"` is a valid Nix expression, it can be [`import`](./builtins.md#builtins-import)ed.
+That requires reading from the output [store path](@docroot@/glossary.md#gloss-store-path) of `drv`, which has to be [realised] before its contents can be read and evaluated.
+
+## Illustration
+
+The following diagram shows how evaluation is interrupted by a build, if the value of a Nix expression depends on realising a store object.
+
+```
++----------------------+             +------------------------+
+| Nix language         |             | Nix store              |
+|  .----------------.  |             |                        |
+|  | Nix expression |  |             |                        |
+|  '----------------'  |             |                        |
+|          |           |             |                        |
+|       evaluate       |             |                        |
+|          |           |             |                        |
+|          V           |             |                        |
+|    .------------.    |             |  .------------------.  |
+|    | derivation |----|-instantiate-|->| store derivation |  |
+|    '------------'    |             |  '------------------'  |
+|                      |             |           |            |
+|                      |             |        realise         |
```

Comment:
If I were to read this as a UML sequence diagram, `realise` would be an internal message presumably triggered by the creation of `store derivation`, whereas in truth instantiation is a number of "add to store" operations, and "realise" is another, independent message that follows the instantiation.
(Also it would take to the store about a derivation, rather than talk to the derivation)

A UML sequence diagram would also highlight that the evaluation process exists during the realisation, whereas this diagram only shows that the Nix language exists. Maybe change the block title from "Nix language" to "Nix evaluator"?

Reply:
I'd be open to pair on fixing the diagram. Admittedly it's a bit hand-wavy but I fear it would get too noisy if done technically correct.

Reply:
Sgtm. Let me start with a proposal, based on my suggestion https://github.com/NixOS/nix/pull/7332#pullrequestreview-1535842400, highlighting that its derivation names aren't great yet:

```
┌───────┐     ┌─────────────┐         ┌─────────┐
│Nix CLI│     │Nix evaluator│         │Nix store│
└───┬───┘     └──────┬──────┘         └────┬────┘
    │                │                     │     
    │evaluate IFD.nix│                     │     
    │───────────────>│                     │     
    │                │                     │     
    │                │instantiate hello.drv│     
    │                │────────────────────>│     
    │                │                     │     
    │                │  realise hello.drv  │     
    │                │────────────────────>│     
    │                │                     │     
    │                │  hello.out = hello  │     
    │                │<────────────────────│     
    │                │                     │     
    │                │instantiate Hello.drv│     
    │                │────────────────────>│     
    │                │                     │     
    │ [ Hello.drv ]  │                     │     
    │<───────────────│                     │     
    │                │                     │     
    │          plan [ Hello.drv ]          │     
    │─────────────────────────────────────>│     
    │                │                     │     
    │        realise [ Hello.drv ]         │     
    │─────────────────────────────────────>│     
┌───┴───┐     ┌──────┴──────┐         ┌────┴────┐
│Nix CLI│     │Nix evaluator│         │Nix store│
└───────┘     └─────────────┘         └─────────┘

```

<details><summary>Source</summary>

https://arthursonzogni.com/Diagon/#Sequence

```
Nix CLI -> Nix evaluator: evaluate IFD.nix

Nix evaluator -> Nix store: instantiate hello.drv
Nix evaluator -> Nix store: realise hello.drv
Nix store -> Nix evaluator: hello.out = hello

Nix evaluator -> Nix store: instantiate Hello.drv
Nix evaluator -> Nix CLI: [ Hello.drv ]

Nix CLI -> Nix store: plan [ Hello.drv ]
Nix CLI -> Nix store: realise [ Hello.drv ]


```

</details>